File: 01.brute-force/solve_py/leetcode_38.py
"""
run length encoding : 문자열 압축 방법. 연속되는 같은 문자가 등장할 때 (개수 + 값) 으로 나타내는 방법
n -> count and say 의 n 번째 원소를 리턴해라
countAndSay(n) = countAndSay(n-1)의 run length encoding 값

재귀적으로 정의됨 -> advanced : 반복문
"""
def count_and_say(n: int) -> str:
    if n == 1:
        return "1"

    ans = ""
    target = count_and_say(n-1)
    count = 0
    prev = target[0]

    for c in target:
        if c != prev:
            ans += str(count) + prev
            prev = c
            count = 1
        else:
            count += 1

    return ans + str(count) + prev

count_and_say(1)
count_and_say(2)
count_and_say(3)
count_and_say(4)
count_and_say(5)
count_and_say(30)

File: 01.brute-force/solve_cpp/leetcode_38.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    string countAndSay(int n) {
        string target = "1";

        for (int i = 1; i < n; i++) {
            string next = "";
            char prev = target[0];
            int count = 0;

            for (int j = 0; j < target.length(); j++) {
                if (target[j] != prev) {
                    next += to_string(count) + prev;
                    prev = target[j];
                    count = 1;
                } else {
                    count += 1;
                }
            }

            target = next + to_string(count) + prev;
        }

        return target;
    }
};

int main() {
    Solution solution;

    cout << "n = 1, " << solution.countAndSay(1) << endl;
    cout << "n = 2, " << solution.countAndSay(2) << endl;
    cout << "n = 3, " << solution.countAndSay(3) << endl;
    cout << "n = 4, " << solution.countAndSay(4) << endl;
    cout << "n = 5, " << solution.countAndSay(5) << endl;
    cout << "n = 30, " << solution.countAndSay(30) << endl;

    return 0;
}